### PR TITLE
[Form] Add input_format option to DateType, DateTimeType and TimeType

### PR DESCRIPTION
**Pull request by fancyweb**
Related Symfony PRs : https://github.com/symfony/symfony/pull/29887 and https://github.com/symfony/symfony/pull/30358

### CODE REVIEW

File: reference/forms/types/options/date_input_format_description.rst.inc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+If the ``input`` option is set to ``string``, this option specifies the format
+of the date. This must be a valid `PHP date format`_.
+
+.. _`PHP date format`: https://secure.php.net/manual/fr/function.date.php
```
Comment (andrew-demb) [inline reference/forms/types/options/date_input_format_description.rst.inc]: Fix locale link FR -> EN
```suggestion
.. _`PHP date format`: https://secure.php.net/manual/en/function.date.php
```

Comment (fancyweb) [reply reference/forms/types/options/date_input_format_description.rst.inc]: Nice catch, thanks !

File: reference/forms/types/datetime.rst [hunk 2]
```diff
@@ -161,6 +162,13 @@ this format.
 
 .. include:: /reference/forms/types/options/seconds.rst.inc
 
+input_format
```
Comment (xabbuh) [inline reference/forms/types/datetime.rst]: I think we should move this description directly below the description of the `input` option.

Comment (fancyweb) [reply reference/forms/types/datetime.rst]: Definitely! That was a mistake.

**Review by xabbuh — APPROVED** [review]
excellent 👍

**Comment by OskarStark** [comment]
The code is merged, so this can be merged too, right?

**Comment by fancyweb** [comment]
@xabbuh Do you want me to add the doc for https://github.com/symfony/symfony/pull/30358 here since it's related and not yet merged?

**Comment by xabbuh** [comment]
@fancyweb Would be nice if you could do that.

**Comment by fancyweb** [comment]
@xabbuh Done!

Status: Needs Review

File: reference/forms/types/options/date_input_format.rst.inc [hunk 3]
```diff
@@ -0,0 +1,6 @@
+input_format
+~~~~~~~~~~~~
+
+**type**: ``string`` **default**: ``Y-m-d``
+
```
Comment (OskarStark) [inline reference/forms/types/options/date_input_format.rst.inc]: Please add a versionadded directive

File: reference/forms/types/datetime.rst [hunk 4]
```diff
@@ -153,6 +154,13 @@ this format.
 
 .. include:: /reference/forms/types/options/_date_limitation.rst.inc
 
+input_format
+~~~~~~~~~~~~
+
+**type**: ``string`` **default**: ``Y-m-d H:i:s``
+
```
Comment (OskarStark) [inline reference/forms/types/datetime.rst]: Please add a versionadded directive

File: reference/forms/types/time.rst [hunk 5]
```diff
@@ -129,6 +130,16 @@ on your underlying object. Valid values are:
 The value that comes back from the form will also be normalized back into
 this format.
 
+input_format
+~~~~~~~~~~~~
+
+**type**: ``string`` **default**: ``H:i:s``
+
```
Comment (OskarStark) [inline reference/forms/types/time.rst]: Please add versionadded directive

**Review by OskarStark — CHANGES_REQUESTED** [review]

**Comment by fancyweb** [comment]
@OskarStark Added the "versionadded" directives at the end of each option block. I had to take a decision because their position in the current documentation is not really consistent.

File: reference/forms/types/options/date_input_format_description.rst.inc [hunk 6]
```diff
@@ -0,0 +1,8 @@
+If the ``input`` option is set to ``string``, this option specifies the format
+of the date. This must be a valid `PHP date format`_.
+
+.. _`PHP date format`: https://secure.php.net/manual/en/function.date.php
+
+.. versionadded:: 4.3
```
Comment (xabbuh) [inline reference/forms/types/options/date_input_format_description.rst.inc]: Sorry if that gives you extra work, but I think we should move this below the `**type**` line everywhere. If there are other places where don't have it like that we should IMO update them in different PRs.

Comment (fancyweb) [reply reference/forms/types/options/date_input_format_description.rst.inc]: Done.

File: reference/forms/types/time.rst [hunk 7]
```diff
@@ -129,6 +130,20 @@ on your underlying object. Valid values are:
...
+.. versionadded:: 4.3
+
+    The ``input_format`` option was introduced in Symfony 4.3.
+
+If the ``input`` option is set to ``string``, this option specifies the format
+of the time. This must be a valid `PHP time format`_.
+
+.. _`PHP time format`: https://secure.php.net/manual/en/function.date.php
```
Comment (xabbuh) [inline reference/forms/types/time.rst]: Can you move this to the end of the file please?

File: reference/forms/types/time.rst [hunk 8]
```diff
@@ -129,6 +130,20 @@ on your underlying object. Valid values are:
...
+input_format
+~~~~~~~~~~~~
+
+**type**: ``string`` **default**: ``H:i:s``
+
+.. versionadded:: 4.3
+
+    The ``input_format`` option was introduced in Symfony 4.3.
+
+If the ``input`` option is set to ``string``, this option specifies the format
```
Comment (xabbuh) [inline reference/forms/types/time.rst]: I think we need to shorten this a bit as you cannot pass anything else then a string:

```rst
The ``input`` option specifies the format of the time. It must be a valid `PHP time format`_.
```

File: reference/forms/types/options/date_input_format_description.rst.inc [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+
+    The ``input_format`` option was introduced in Symfony 4.3.
+
+If the ``input`` option is set to ``string``, this option specifies the format
```
Comment (xabbuh) [inline reference/forms/types/options/date_input_format_description.rst.inc]: I think we need to shorten this a bit as you cannot pass anything else then a string:

```rst
The ``input`` option specifies the format of the date. It must be a valid `PHP time format`_.
```

**Comment by xabbuh** [comment]
I have left some last minor comments. After that we are good. Thank you for the great work! 👍

File: reference/forms/types/options/date_input_format_description.rst.inc (hunk 9 above)
Comment (fancyweb) [reply reference/forms/types/options/date_input_format_description.rst.inc]: 🤔 The `input_format` option is only useful if the `input` option is set to `string`. We can put something else in the `input` option. It seems logic to me to warn the user about this.

This is how it is done in many other places (eg : If the ``widget`` option is set to ``single_text``, this option specifies the format of the input.)

Comment (xabbuh) [reply reference/forms/types/options/date_input_format_description.rst.inc]: Oh I am sorry, I misread that the first sentence is talking about the `input` option (and not the `input_format` option.

**Review by xabbuh — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]
Thank you, great job 👍🏻

**Comment by javiereguiluz** [comment]
Thanks a lot Thomas!

**Pull request closed without merging**